### ISSUE TITLE
Define handling of multiple applicable modifiers

### ISSUE THREAD

**Issue by mnoorenberghe** [w3c/browser-payment-api#684]
1) From what I can tell, the spec doesn't currently define which `additionalDisplayItems` should be shown if a UA decides to show any.
2) More generally, handling of multiple applicable modifiers isn't clear. It looks like it's possible to have multiple `methodData` and `modifiers` with the same `supportedMethods` including the same `data`.

Consider the following example snippet:
```js
// Credit card incurs a $3.00 processing fee.
const creditCardFee = {
  label: "Credit card processing fee",
  amount: { currency: "USD", value: "3.00" },
};

// Visa card incurs a $1.00 processing fee.
const visaFee = {
  label: "Visa processing fee",
  amount: { currency: "USD", value: "1.00" },
};

// Modifiers apply when the user chooses to pay with
// a credit card.
const modifiers = [
  {
    additionalDisplayItems: [creditCardFee],
    supportedMethods: "basic-card",
    total: {
      label: "Total due",
      amount: { currency: "USD", value: "53.00" },
    },
    data: {
      supportedTypes: ["credit"],
    },
  },
  {
    additionalDisplayItems: [visaFee],
    supportedMethods: "basic-card",
    total: {
      label: "Total due",
      amount: { currency: "USD", value: "54.00" },
    },
    data: {
      supportedNetworks: ["visa"]
    },
  },
];
Object.assign(details, { modifiers });
```
If multiple modifiers match the selected payment method (e.g. if the user selects a Visa card):
* Should all applicable additional display items be shown? The first matching one? Last matching one? Most specific one (would need to be defined)
* Which modifier's `total` should be used?

This seems like something that should be tested as well.

**Comment by ianbjacobs** [comment]
@mnoorenberghe, here's my recollection:

1) This is issue 309:
    https://github.com/w3c/payment-request/issues/309

 2) We discussed in March 2017:
     https://www.w3.org/2017/03/23-wpwg-minutes

 3) I believe the "first match wins" approach was adopted
     https://github.com/w3c/payment-request/commit/2972fbb86df49ef7a4f4b9bdfc8b5e3c1c1aee31

@marcoscaceres, there was text in that pull request to cover this question. I don't see it in the latest draft of the spec. Is it implicit in one of the algorithms now?

Thank you,
Ian

**Comment by marcoscaceres** [comment]
I was also sure there was spec text covering this case. Will have a look tomorrow.

**Comment by marcoscaceres** [comment]
Need to look at https://github.com/w3c/payment-request/pull/581 which closed #309. It includes a test.

**Comment by mnoorenberghe** [comment]
Not sure how I didn't find #309 in my search… It looks like duplicate PMIs were disallowed while working on #581 (which I don't think is a good solution and it seemed others agree) but that text was removed in https://github.com/w3c/payment-request/pull/581/commits/5e6ca6da68bd65634a309662c067c3b83eed97fc before merging. It seems like #581 then didn't end up addressing the issue.

There was some talk about having the first match win but I'll note that shippingOptions work in the opposite way… the last `selected` wins: 

> If more than one item in the sequence has selected set to true, then the user agent selects the last one in the sequence. 

I think we should be consistent on first vs. last in tie breaker scenarios.

**Comment by marcoscaceres** [comment]
>  It seems like #581 then didn't end up addressing the issue.

I think you are right... seems we talked ourselves out of a proper solution in #581 or somehow the solution got lost. 

> I think we should be consistent on first vs. last in tie breaker scenarios.

Agree. From my testing Chrome's behaviour seems to be "**only the last matching PMI declaration wins**  - all previous ones are ignored". @rsolomakhin, @zkoch - please confirm... as this might confuse developers, leading potentially to incorrect totals 💸!  

To be somewhat sane, we might need an algorithm that, in order, filter `supportedMethods` on PMI, then filter `supportedTypes`, and finally filter on `supportedNetworks` (need to code it up to verify).  

Developers then need to go specify things from least specific "All the [debit, credit, prepay] cards incur a $3 fee!" to least specific, "Visa credit cards $2 fee". 

< 📣 rant 📣 >
I'll admit, this makes me queasy 🤢. 

The right way to deal with this would have been to have a payment option change event, that includes just `"{type: 'credit', network: 'visa'}` (i.e., we don't leak private info or credit card numbers!). Then the developer can add additional display items, update the total via`.updateWith()`... like we do with shipping change events. 

Why don't we just do that instead? Better than dealing with insane combinations of modifiers and vastly simplifies the spec for both implementers, developers, and reduces the risk of screwups for everyone.  
</📣 rant 📣>

**Comment by marcoscaceres** [comment]
Still feeling a little rambly, so this would roughly look like:

```JS
const creditCardFees = new Map([
  //Cards, as per https://www.w3.org/Payments/card-network-ids
  ["visa", { currency: "USD", value: "3.0" }],
]);

const debitCardFees = new Map([
  ["visa", { currency: "USD", value: "2.0" }],
]);

const fees = new Map([
  // BasicCardType enum values
  ["debit", debitCardFees]
  ["credit", creditCardFees],
  ["prepaid", creditCardFees],
]);

request.onpaymentoptionchange = function feeFinder(ev) {
  const { type, network } = ev.details;
  // look up fee, or standard fee
  const fee = fees.get(type).get(network) || { currency: "USD", value: "5.00" };
  const newTotal = updateTotal(oldTotal, fee);
  const additonalDisplayItems = [{ label: `${type} card fee`, amount: fee };
  ev.updateWith({ newTotal, additonalDisplayItems });
});
```

As you can see, the above doesn't require any complex modifier logic, just a  super simple lookup table.

**Comment by adrianhopebailie** [comment]
@marcoscaceres I'm assuming this would need a slight tweak

```javascript
request.onpaymentoptionchange = function feeFinder(ev) {
  if(ev.method == "basic-card"){
    // as above...
  }
  if(ev.method == "https://android.com/pay") {
    // do other stuff
  }
});
```

This would mean that each payment method must define which data in a payment instrument is private so that it is not returned in the event.

Not insurmountable but likely more challenging for the open payment methods? They'd need to define this in their manifest somehow.

@rsolomakhin WDYT?

**Comment by rsolomakhin** [comment]
@adrianhopebailie : What's your definition of an open payment method?

I would think that payment methods implemented inside of the browser (so far only `"basic-card"`) would let the merchant know the selected instrument properties (type and network for `"basic-card"`), because the selection happens inside of the browser payment sheet. For payment methods implemented in a payment app, Chrome currently does not have a way to know which instrument is being selected inside of the payment app, so Chrome could let the merchant know only the list of payment methods supported by the selected payment app. From spec perspective, we could make `PaymentOptionChangeEvent.method` a union of string and array and suggest that browsers populate `PaymentOptionChangeEvent.details` to the best of their ability.

**Comment by rsolomakhin** [comment]
@marcoscaceres: We should be selecting the first modifier on both [Android](https://cs.chromium.org/chromium/src/chrome/android/java/src/org/chromium/chrome/browser/payments/PaymentRequestImpl.java?rcl=c67eed882990b58d681e7ff19389a3363660f873&l=961) and [desktop](https://cs.chromium.org/chromium/src/components/payments/content/payment_request_spec.cc?rcl=b323d65639ba99b7c8b191a45018c11b3a022239&l=248).

Thank you, @anthonyvd and @gogerald, for looking this up so quickly!

**Comment by mnoorenberghe** [comment]
> The right way to deal with this would have been to have a payment option change event, that includes just "{type: 'credit', network: 'visa'} (i.e., we don't leak private info or credit card numbers!). Then the developer can add additional display items, update the total via.updateWith()... like we do with shipping change events.

This sounds like a great idea and I'd be very much in favour of this.

**Comment by marcoscaceres** [comment]
@adrianhopebailie, you are absolutely correct. 

This seems to align with what ApplePay already does and with what we had discussed with @aestes in https://github.com/w3c/payment-request/issues/662#issuecomment-359038775 where we would add the payment method to the `PaymentChangeEvent` interface. 

Discussing with people at Mozilla, we think this proposal (let's get rid of modifiers) warrants serious consideration. It solves serious problems in the specification, and greatly simplifies the overall design with no loss in functionality. 

@adrianba, @zkoch, @aestes, @domenic, thoughts? 

cc @stpeter.

**Comment by adrianhopebailie** [comment]
@marcoscaceres If I recall correctly we avoided this before because we felt there was no way to be certain that the payment handler (if it wasn't the browser) would be able to emit such an event. I.e. IPC between the handler (if it is a mobile app for example) and the browser is not guaranteed to be available.

I think that based on where we are with the handler architecture that view has probably softened and this would be the ideal design. I'm definitely a +1

@rsolomakhin asked:

> What's your definition of an open payment method?

Those with a URL payment method identifier. For those payment methods the browser doesn't know which fields to redact if it was to emit a `PaymentChangeEvent`.

> For payment methods implemented in a payment app, Chrome currently does not have a way to know which instrument is being selected inside of the payment app, so Chrome could let the merchant know only the list of payment methods supported by the selected payment app.

That sounds like something we want to add to Payment Handler API? The ability to emit such an event. In this case it would be up to the handler to ensure it doesn't leak private data.

**Comment by stpeter** [comment]
@marcoscaceres +1 ... Reducing (seemingly unnecessary) complexity seems like a win here.

**Comment by rsolomakhin** [comment]
Mistake!

**Comment by rsolomakhin** [comment]
One caveat to consider in favor of keeping modifiers: They allow browsers to show the modified price in the payment sheet, so the user can glance over all options and make a better decision that way. Browsers can also use the different prices as a signal when sorting the payment instruments and pre-selecting the default one. See the mastercard discount in the screenshot below, for example:

<img src="https://i.imgur.com/6Lx8Kor.png" width="400">

**Comment by rsolomakhin** [comment]
@madmath cc

**Comment by marcoscaceres** [comment]
Tl;dr: this is a Payment Handler specific issue. 

Looking at the history of this bug, related discussions, etc. This is what needs to happen (I think):

 * The Payment Request spec just gathers Payment Method Modifier tuples and then passes them to the user's selected Payment Handler.  
 * Payment Handler (e.g., basic card) then needs to make a decision about how to honor the modifiers. That is, in order, it can look at what the user has picked, and run down the list matching from least specific, to most specific. 

Thus, the Payment Request spec can't help here, because it doesn't know anything about the `.data`. 

Best we can do in Payment Request is suggest people consult the payment handler for matching order.

**Comment by stpeter** [comment]
That does seem to be the best we can do. (And it doesn't even address cross-method interactions - save X with Interledger, Y with SEPA, Z with debit - yikes!)

**Comment by ianbjacobs** [comment]
I am hearing you say:

 * How duplicates are handled is payment method specific. This implies that third party payment handlers will need access to all of the PMI/data tuples (including duplicates) in order to respond to canMakePayment().
 * Basic Card is expected to define an algorithm for handling duplicates, with a "last one wins"  approach when all modifiers are equivalent.
 * The PR API spec will explain that handling duplicates is payment method specific and for consistency with basic card, new payment methods should define a "last one wins" approach.

Ian

**Comment by ianbjacobs** [comment]
I've raised https://github.com/w3c/payment-method-basic-card/issues/60
so that we can close the current issue more quickly with an informative edit to PR API.
